litejpeg/core/huffman/huffmancore.py

```diff
@@ -0,0 +1,479 @@
+are required for the process of encoding the input data.
+This module takes input from the RLE module containing three paramenters:
+1. amplitude : Non-zero input in the input matrix of the RLE.
+2. runlength : Number of zeros before the non-zero amplitude.
+3. size : Number of bits required to store amplitude.
+Each of the output is of 8 bits.
+The main purpose of the module is to serialize the data and encode the data
+using Huffman tables.
+The advantages of the encoding scheme is that boundaries between characters
+are easily determined, and the pattern used for each character is fixed and universal.
+
+Input data format : 20 bits
+input_data[0:12] = amplitude
+input_data[12:16] = Size
+input_data[16:20] = Runlength
+
+Output data format: 8 bits
+output_data[0:8]
+"""
+
+from litex.gen import *
+from litex.soc.interconnect.stream import *
+
+from litejpeg.core.common import *
+from litejpeg.core.huffman.ac_rom import ac_rom_core
+from litejpeg.core.huffman.dc_rom import dc_rom_core
+
+
+# This is introduced to provide delay in the module for stablizing the
+# input coming from the other modules.
+# This provides a delay of 2 clock cycles.
+datapath_latency = 2
+
+
+# Provide an additional clock in the module for synchronizing the input and
+# output data.
+@CEInserter()
+class HuffmanDatapath(Module):
+    """ This gives the datapath for the serialization of the input data.
+    The data coming to the module is been controlled by using
+    ready_data_write and ready_data_read, as when the module is ready
+    then the input ready_data_write will be 1 and when the data to the
+    output is ready then ready_data_read is 1.
+
+    Parameters:
+    ----------
+    amplitude : 12 bits
+                Non-zero amplitudes in the matrix.
+    size      : 4 bits
+                Number of bits required to store the amplitude.
+    runlength : 4 bits
+                Number of zeros before the non-zero amplitude.
+    word_count: 6 bits
+                Display the index of the matrix.
+    source    : 8 bits
+                Output of the serial data coming out.
+    """
+    def __init__(self):
+
+        # Getting input from the HuffmanEncoder module.
+        self.amplitude = amplitude = Signal(12)
+        self.size = size = Signal(4)
+        self.runlength = runlength = Signal(4)
+        self.word_count = Signal(6)
+        # Giving the output to the HuffmanEncoder module.
+        self.source = source = Record(block_layout(9))
+
+        # Getting ready the input and output data.
+
+        # See wheather the input is ready to be process.
+        self.ready_data_write = Signal(1)
+        # See wheather the output is ready to process.
+        self.ready_data_read = Signal(1)
+
+        # Attaching the AC and DC ROM to get the Encoded values in order to get
+        # the encoded values.
+        self.submodules.dc_rom_got = dc_rom_core()
+        self.submodules.ac_rom_got = ac_rom_core()
```

Comment:
`dc_rom_got` ? What does the `got` mean? Should this just be `self.submodules.dc_rom`?

Reply:
Done